kumascript/macros/InheritanceDiagram.ejs

```diff
@@ -1,132 +1,120 @@
-    x -= 10;
-    str += '<line x1="'+x+'" y1="'+(y+24)+'" x2="'+(x-30)+'" y2="'+(y+24)+'" stroke="#D4DDE4"/>';
-  } else {
-    str += '<polyline points="'+x+','+(y+24)+'  '+(x+10)+','+(y+19)+'  '+(x+10)+','+(y+29)+'  '+x+','+(y+24)+'" stroke="#D4DDE4" fill="none"/>';
-    x += 10;
-    str += '<line x1="'+x+'" y1="'+(y+24)+'" x2="'+(x+30)+'" y2="'+(y+24)+'" stroke="#D4DDE4"/>';
+    polylinePoints = `${x},${y+14} ${x-10},${y+9} ${x-10},${y+19} ${x},${y+14}`;
+    linePoints = `x1="${x-10}" y1="${y+14}" x2="${x-30}" y2="${y+14}"`;
   }
+
+  str += `<polyline points="${polylinePoints}" stroke="#D4DDE4" fill="none"/>`;
+  x -= 10;
+  str += `<line ${linePoints} stroke="#D4DDE4"/>`;
+
   return str;
 }
 
-var padding = (height / width) * 100;
-var xpos = 1;
-var ypos = 1;
-var viewbox = 'viewbox="-'+ offset + ' 0 '+ width + ' ' + height +'"';
-
-if (rtl) {
-  viewbox = 'viewbox="510 0 '+ width + ' ' + height +'" transform="scale(-1 1)"';
+function connectingLineWithTriangle() {
+  out += `<polyline points="${xpos},${ypos+14} ${xpos+10},${ypos+9} ${xpos+10},${ypos+19} ${xpos},${ypos+14}" stroke="#D4DDE4" fill="none"/>`;
+  xpos += 10;
+  out += `<line x1="${xpos}" y1="${ypos+14}" x2="${xpos+8}" y2="${ypos+14}" stroke="#D4DDE4"/>`;
+  xpos += 8;
+  out += `<line x1="${xpos}" y1="${ypos+14}" x2="${xpos}" y2="${ypos+59}" stroke="#D4DDE4"/>`;
+  ypos = 60;
+  out += `<line x1="${xpos}" y1="${ypos}" x2="${xpos-17}" y2="${ypos}" stroke="#D4DDE4"/>`;
+  xpos -= 18;
+  ypos = 47
 }
-result = '<div id="interfaceDiagram" style="display: inline-block; position: relative; width: 100%; padding-bottom: '+ padding +'%; vertical-align: middle; overflow: hidden;">' +
-             '<svg style="display: inline-block; position: absolute; top: 0; left: 0;" '+ viewbox +' preserveAspectRatio="xMinYMin meet">';
-
-for (var i = 0; i < inheritanceChain.length; i++) {
-  var fill = (inheritanceChain[i] == mainIF) ? '#F4F7F8': '#fff';
-  // we are going to the second row with a special line
-  if (i == 4) {
-    result += '<polyline points="'+xpos+','+(ypos+24)+'  '+(xpos+10)+','+(ypos+19)+'  '+(xpos+10)+','+(ypos+29)+'  '+xpos+','+(ypos+24)+'" stroke="#D4DDE4" fill="none"/>'
-    xpos += 10;
-    result += '<line x1="'+xpos+'" y1="'+(ypos+24)+'" x2="'+(xpos+8)+'" y2="'+(ypos+24)+'" stroke="#D4DDE4"/>';
-    xpos += 8;
-    result += '<line x1="'+xpos+'" y1="'+(ypos+24)+'" x2="'+xpos+'" y2="'+(ypos+89)+'" stroke="#D4DDE4"/>';
-    ypos = 90;
-    result += '<line x1="'+xpos+'" y1="'+ypos+'" x2="'+(xpos-17)+'" y2="'+ypos+'" stroke="#D4DDE4"/>';
-    xpos -= 18;
-    ypos = 65;
-  }
-  // first row
-  if (i < 4) {
-    result += rectWithText(xpos, ypos, fill, inheritanceChain[i]);
-    if (i != inheritanceChain.length - 1 && i != 3) {
-      result += lineWithTriangle(xpos, ypos);
-      xpos += 40;
+
+function drawDiagram(inheritanceChain) {
```

Comment:
I don't like the way this function works by modifying the global `out`. It's like programming by side effects:

```js
let out = '';
...
drawDiagram(inheritanceChain);
// do something with out because it's different now
```

...rather than:

```js
let out = drawDiagram(inheritanceChain);
// do something with out
```

Some of these functions (like `lineWithTriangle`) do follow a pattern of returning the string, but others, like this one and `connectingLineWithTriangle`, don't. It would be good to be consistent.